### PR TITLE
fix(Linker): Don't use vendor to check bill type

### PR DESCRIPTION
**Pull request by drazik**
Until now, we maintained a vendors list to check if a bill has the
health type or not. This is not very scalable because we have to add new
vendors when new connectors are released, and we can do typos. Using the
bill `type` property is much more reliable.

### CODE REVIEW

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.js [hunk 1]
```diff
@@ -45,7 +31,7 @@ const isUncategorizedOperation = operation => {
 }
 
 const isHealthBill = bill => {
-  return includes(HEALTH_VENDORS, bill.vendor)
+  return bill.type === 'health_costs'
```
Comment (ptbrowne) [inline src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.js]: Are you sure there is no risk to have false negatives here ? I mean that we must be sure that every vendor here correctly sets the bill.type. Shouldn't we do 

```
  return  bill.type === 'health_costs' || includes(HEALTH_VENDORS, bill.vendor)
```

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js [hunk 2]
```diff
@@ -31,9 +31,9 @@ describe('operations filters', () => {
       const bill = { matchingCriterias: { labelRegex: 'Bidule' } }
       const fByBrand = filterByBrand(bill)
 
-      expect(fByBrand({ label: 'Chez Bidule !!!' })).toBe(true)
-      expect(fByBrand({ label: 'bidule 20/08' })).toBe(true)
-      expect(fByBrand({ label: 'Ca match paaaas du tout' })).toBe(false)
+      expect(fByBrand({ label: 'Chez Bidule !!!' })).toBeTruthy()
+      expect(fByBrand({ label: 'bidule 20/08' })).toBeTruthy()
+      expect(fByBrand({ label: 'Ca match paaaas du tout' })).toBeFalsy()
```
Comment (ptbrowne) [inline src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: Truthy and falsy are not the same as true/false, it seems we are lowering the bar of the test here, what is the gain ?

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js [hunk 3]
```diff
@@ -78,8 +78,8 @@ describe('operations filters', () => {
   const HEALTH_INSURANCE_CAT = '400620'
 
   describe('filtering by category', () => {
-    fit('should only match bills with the right categoryId when the vendor is known to be a health insurance provider', () => {
-      const fByCategory = filterByCategory({ vendor: 'Ameli' })
+    it('should only match bills with the right categoryId when the bill type is health_costs', () => {
```
Comment (ptbrowne) [inline src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: fit -> it. oups :)

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js [hunk 4]
```diff
@@ -140,14 +140,9 @@ describe('operations filters', () => {
...
       const fByCategory = filterByCategory({ vendor: 'SFR' })
       expect(fByCategory({ manualCategoryId: HEALTH_EXPENSE_CAT })).toBeFalsy()
-      expect(
-        fByCategory({ automaticCategoryId: HEALTH_EXPENSE_CAT })
-      ).toBeFalsy()
       expect(fByCategory({ manualCategoryId: '0' })).toBeTruthy()
-      expect(fByCategory({ automaticCategoryId: '0' })).toBeTruthy()
       expect(fByCategory({})).toBeTruthy()
       expect(fByCategory({ manualCategoryId: '400611' })).toBeTruthy()
-      expect(fByCategory({ automaticCategoryId: '400611' })).toBeTruthy()
```
Comment (ptbrowne) [inline src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: For my information, why did you remove those lines ?

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.js (hunk 1 above)
Comment (drazik) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.js]: @doubleface checked that all health connectors sets the `bill.type` to `health_costs`, so I think it is safe

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js [hunk 5]
```diff
@@ -177,14 +172,14 @@ describe('operations filters', () => {
         label: 'Visite chez le médecin',
         _id: 'o1',
         date: new Date(2017, 11, 13),
-        automaticCategoryId: '400610'
+        manualCategoryId: '400610'
```
Comment (ptbrowne) [inline src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: Question: why do we change from automatic to manual.

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js (hunk 4 above)
Comment (drazik) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: Because the case in which we only have `automaticCategoryId` means that our categorization models have not been applied, and in that case the `getCategoryId` helper returns `null`. That's how we represent the "categorization pending" state when a transaction have been improted but not yet categorized by our categorization service.

In tests, I think it's better to use only `manualCategoryId` because when it is present, it is always used. So we don't have to think about all the specific cases implemented by `getCategoryId` in every test that uses categories. By always using `manualCategoryId`, we are certain that this category will be used.

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js (hunk 5 above)
Comment (drazik) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: See https://github.com/cozy/cozy-banks/pull/1498#discussion_r317960624

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js (hunk 2 above)
Comment (drazik) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: Actually I don't know why but these tests were passing and then they were failing. I don't know how they passed before since `fByBrand` returns the result of `label.match(regex)`, so it returns `null` or an array, but never a boolean. That's why I  used `toBeTruthy` and `toBeFalsy` to fix this. But another way would be to `return Boolean(label.match(regex))` and still use `toBe(false)` and `toBe(true)` in tests. What do you think?

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js (hunk 3 above)
Comment (drazik) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: This is on purpose

File: src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js (hunk 2 above)
Comment (ptbrowne) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: I think it is better to be fairly strict on what we return, so Boolean(label.match(regex)) is good to me.

> Be conservative in what you send, be liberal in what you accept

https://en.wikipedia.org/wiki/Robustness_principle

Comment (ptbrowne) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: Otherwise, at some point, someones sees that you return the regex result, and starts using it instead of considering it as a boolean, then you want to change the behavior and not use a regex engine, and you break the implicit API for the other person.

Comment (drazik) [reply src/ducks/billsMatching/Linker/billsToOperation/operationsFilters.spec.js]: :+1:
